### PR TITLE
PYTHON-3032 Add support for Python 3.10 and pymongo 3.12

### LINKED ISSUE #14: Update stubs for PyMongo 3.12
```
Run stubtest bson pymongo gridfs --ignore-missing-stub --whitelist tests/stubtest_whitelist/common.txt
error: pymongo.MongoClient.start_session is inconsistent, runtime argument "causal_consistency" has a default value of type None, which is incompatible with stub argument type builtins.bool
Stub: at line 89
def (self: pymongo.mongo_client.MongoClient, causal_consistency: builtins.bool =, default_transaction_options: Union[pymongo.client_session.TransactionOptions, None] =) -> pymongo.client_session.ClientSession
Runtime: at line 1888 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/mongo_client.py
def (self, causal_consistency=None, default_transaction_options=None, snapshot=False)

error: pymongo.MongoClient.start_session is inconsistent, stub does not have argument "snapshot"
Stub: at line 89
def (self: pymongo.mongo_client.MongoClient, causal_consistency: builtins.bool =, default_transaction_options: Union[pymongo.client_session.TransactionOptions, None] =) -> pymongo.client_session.ClientSession
Runtime: at line 1888 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/mongo_client.py
def (self, causal_consistency=None, default_transaction_options=None, snapshot=False)

error: pymongo.client_session.SessionOptions.__init__ is inconsistent, runtime argument "causal_consistency" has a default value of type None, which is incompatible with stub argument type builtins.bool
Stub: at line 10
def (self: pymongo.client_session.SessionOptions, causal_consistency: builtins.bool =, default_transaction_options: Union[pymongo.client_session.TransactionOptions, None] =)
Runtime: at line 176 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/client_session.py
def (self, causal_consistency=None, default_transaction_options=None, snapshot=False)

error: pymongo.client_session.SessionOptions.__init__ is inconsistent, stub does not have argument "snapshot"
Stub: at line 10
def (self: pymongo.client_session.SessionOptions, causal_consistency: builtins.bool =, default_transaction_options: Union[pymongo.client_session.TransactionOptions, None] =)
Runtime: at line 176 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/client_session.py
def (self, causal_consistency=None, default_transaction_options=None, snapshot=False)

error: pymongo.collection.Collection.aggregate_raw_batches is inconsistent, stub does not have argument "session"
Stub: at line 146
def (self: pymongo.collection.Collection, pipeline: builtins.list[typing.Mapping[builtins.str, Any]], **kwargs: Any) -> pymongo.cursor.RawBatchCursor
Runtime: at line 2503 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/collection.py
def (self, pipeline, session=None, **kwargs)

error: pymongo.database.Database.set_profiling_level is inconsistent, stub does not have argument "sample_rate"
Stub: at line 120
def (self: pymongo.database.Database, level: builtins.int, slow_ms: Union[builtins.int, None] =, session: Union[pymongo.client_session.ClientSession, None] =)
Runtime: at line 1115 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/database.py
def (self, level, slow_ms=None, session=None, sample_rate=None, filter=None)

error: pymongo.database.Database.set_profiling_level is inconsistent, stub does not have argument "filter"
Stub: at line 120
def (self: pymongo.database.Database, level: builtins.int, slow_ms: Union[builtins.int, None] =, session: Union[pymongo.client_session.ClientSession, None] =)
Runtime: at line 1115 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/database.py
def (self, level, slow_ms=None, session=None, sample_rate=None, filter=None)

error: pymongo.mongo_client.MongoClient.start_session is inconsistent, runtime argument "causal_consistency" has a default value of type None, which is incompatible with stub argument type builtins.bool
Stub: at line 89
def (self: pymongo.mongo_client.MongoClient, causal_consistency: builtins.bool =, default_transaction_options: Union[pymongo.client_session.TransactionOptions, None] =) -> pymongo.client_session.ClientSession
Runtime: at line 1888 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/mongo_client.py
def (self, causal_consistency=None, default_transaction_options=None, snapshot=False)

error: pymongo.mongo_client.MongoClient.start_session is inconsistent, stub does not have argument "snapshot"
Stub: at line 89
def (self: pymongo.mongo_client.MongoClient, causal_consistency: builtins.bool =, default_transaction_options: Union[pymongo.client_session.TransactionOptions, None] =) -> pymongo.client_session.ClientSession
Runtime: at line 1888 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/mongo_client.py
def (self, causal_consistency=None, default_transaction_options=None, snapshot=False)

error: pymongo.monitoring.CommandFailedEvent.__init__ is inconsistent, stub does not have argument "command_name"
Stub: at line 82
def (self: pymongo.monitoring.CommandFailedEvent, duration: datetime.timedelta, failure: typing.Mapping[builtins.str, Any], *args: Any)
Runtime: at line 666 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, failure, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandFailedEvent.__init__ is inconsistent, stub does not have argument "request_id"
Stub: at line 82
def (self: pymongo.monitoring.CommandFailedEvent, duration: datetime.timedelta, failure: typing.Mapping[builtins.str, Any], *args: Any)
Runtime: at line 666 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, failure, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandFailedEvent.__init__ is inconsistent, stub does not have argument "connection_id"
Stub: at line 82
def (self: pymongo.monitoring.CommandFailedEvent, duration: datetime.timedelta, failure: typing.Mapping[builtins.str, Any], *args: Any)
Runtime: at line 666 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, failure, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandFailedEvent.__init__ is inconsistent, stub does not have argument "operation_id"
Stub: at line 82
def (self: pymongo.monitoring.CommandFailedEvent, duration: datetime.timedelta, failure: typing.Mapping[builtins.str, Any], *args: Any)
Runtime: at line 666 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, failure, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandFailedEvent.__init__ is inconsistent, stub does not have argument "service_id"
Stub: at line 82
def (self: pymongo.monitoring.CommandFailedEvent, duration: datetime.timedelta, failure: typing.Mapping[builtins.str, Any], *args: Any)
Runtime: at line 666 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, failure, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandFailedEvent.__init__ is inconsistent, runtime does not have *args argument "args"
Stub: at line 82
def (self: pymongo.monitoring.CommandFailedEvent, duration: datetime.timedelta, failure: typing.Mapping[builtins.str, Any], *args: Any)
Runtime: at line 666 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, failure, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandStartedEvent.__init__ is inconsistent, stub does not have argument "request_id"
Stub: at line 60
def (self: pymongo.monitoring.CommandStartedEvent, command: typing.Mapping[builtins.str, Any], database_name: builtins.str, *args: Any)
Runtime: at line 568 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, command, database_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandStartedEvent.__init__ is inconsistent, stub does not have argument "connection_id"
Stub: at line 60
def (self: pymongo.monitoring.CommandStartedEvent, command: typing.Mapping[builtins.str, Any], database_name: builtins.str, *args: Any)
Runtime: at line 568 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, command, database_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandStartedEvent.__init__ is inconsistent, stub does not have argument "operation_id"
Stub: at line 60
def (self: pymongo.monitoring.CommandStartedEvent, command: typing.Mapping[builtins.str, Any], database_name: builtins.str, *args: Any)
Runtime: at line 568 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, command, database_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandStartedEvent.__init__ is inconsistent, stub does not have argument "service_id"
Stub: at line 60
def (self: pymongo.monitoring.CommandStartedEvent, command: typing.Mapping[builtins.str, Any], database_name: builtins.str, *args: Any)
Runtime: at line 568 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, command, database_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandStartedEvent.__init__ is inconsistent, runtime does not have *args argument "args"
Stub: at line 60
def (self: pymongo.monitoring.CommandStartedEvent, command: typing.Mapping[builtins.str, Any], database_name: builtins.str, *args: Any)
Runtime: at line 568 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, command, database_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.CommandSucceededEvent.__init__ is inconsistent, stub does not have argument "service_id"
Stub: at line 67
def (self: pymongo.monitoring.CommandSucceededEvent, duration: datetime.timedelta, reply: typing.Mapping[builtins.str, Any], command_name: builtins.str, request_id: builtins.int, connection_id: Tuple[builtins.str, Union[builtins.int, None]], operation_id: Union[builtins.int, None])
Runtime: at line 619 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, duration, reply, command_name, request_id, connection_id, operation_id, service_id=None)

error: pymongo.monitoring.PoolClearedEvent.__init__ is inconsistent, stub does not have argument "service_id"
Stub: at line 89
def (self: pymongo.monitoring._PoolEvent, address: Tuple[builtins.str, Union[builtins.int, None]])
Runtime: at line 749 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, address, service_id=None)

error: pymongo.monitoring._CommandEvent.__init__ is inconsistent, stub does not have argument "service_id"
Stub: at line 49
def (self: pymongo.monitoring._CommandEvent, command_name: builtins.str, request_id: builtins.int, connection_id: Tuple[builtins.str, Union[builtins.int, None]], operation_id: Union[builtins.int, None])
Runtime: at line 517 in file /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pymongo/monitoring.py
def (self, command_name, request_id, connection_id, operation_id, service_id=None)
```
https://github.com/ShaneHarvey/pymongo-stubs/runs/3402394195?check_suite_focus=true#step:5:1

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #14

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
I'm confused why black and flake8 code checks are failing. Why do we need to reformat the code when nothing has changed? Do new versions of black/flake8 change the code formatting rules?

**Comment by blink1073** [comment]
> Do new versions of black/flake8 change the code formatting rules?

Flake8 was my fault.  iirc black can change on a major version change.

**Comment by blink1073** [comment]
Another possibility is that black was only run on changed files previously, whereas I just ran it on all files locally.

**Comment by ShaneHarvey** [comment]
I'm even more confused now because we actually pin to exact versions of flake8+black: https://github.com/mongodb-labs/pymongo-stubs/blob/main/requirements-tests-py3.txt

We also run these checkers on all files: https://github.com/mongodb-labs/pymongo-stubs/blob/dcc998b/.github/workflows/tests.yml#L26

Do you mind looking more into this more? I was planning to add black to pymongo too but I wouldn't want to deal with this issue.

**Comment by ShaneHarvey** [comment]
Oh sorry, I didn't realize that you had updated the stubs first.... Ignore everything I said.

File: pymongo-stubs/cursor.pyi [hunk 1]
```diff
@@ -38,9 +38,9 @@ class Cursor:
         max_time_ms: Optional[int] = ...,
         max: Optional[_Sort] = ...,
         min: Optional[_Sort] = ...,
-        return_key: bool = ...,
-        show_record_id: bool = ...,
-        snapshot: bool = ...,
+        return_key: Union[bool, None] = ...,
+        show_record_id: Union[bool, None] = ...,
+        snapshot: Union[bool, None] = ...,
```
Comment (ShaneHarvey) [inline pymongo-stubs/cursor.pyi:43]: `Optional[bool]`

File: pymongo-stubs/client_session.pyi [hunk 2]
```diff
@@ -8,7 +8,10 @@ from pymongo.write_concern import WriteConcern
 
 class SessionOptions:
     def __init__(
-        self, causal_consistency: bool = ..., default_transaction_options: Optional[TransactionOptions] = ...
+        self,
+        causal_consistency: Union[bool, None] = ...,
```
Comment (ShaneHarvey) [inline pymongo-stubs/client_session.pyi:12]: This should use `Optional[bool]` which is a more readable alias of `Union[bool, None]`.

File: pymongo-stubs/mongo_client.pyi [hunk 3]
```diff
@@ -87,7 +87,10 @@ class MongoClient(common.BaseObject):
     def close_cursor(self, cursor_id: int, address: Optional[Tuple[str, Optional[int]]] = ...) -> None: ...
     def kill_cursors(self, cursor_ids: Sequence[int], address: Optional[Tuple[str, Optional[int]]] = ...) -> None: ...
     def start_session(
-        self, causal_consistency: bool = ..., default_transaction_options: Optional[TransactionOptions] = ...
+        self,
+        causal_consistency: Union[bool, None] = ...,
```
Comment (ShaneHarvey) [inline pymongo-stubs/mongo_client.pyi:91]: `Optional[bool]`

File: pymongo-stubs/monitoring.pyi [hunk 4]
```diff
@@ -46,7 +46,14 @@ _Document = Mapping[str, Any]
...
 
 class _CommandEvent:
-    def __init__(self, command_name: str, request_id: int, connection_id: _Address, operation_id: Optional[int]) -> None: ...
+    def __init__(
+        self,
+        command_name: str,
+        request_id: int,
+        connection_id: _Address,
+        operation_id: Optional[int],
+        service_id: Optional[int] = ...,
```
Comment (ShaneHarvey) [inline pymongo-stubs/monitoring.pyi:55]: service_id is an ObjectId so this and everywhere else should be `Optional[ObjectId]`

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Here's my pre commit hook. You're right it only runs on changed files:
```
cat .git/hooks/pre-commit
#!/bin/sh
#
# An example hook script that will run flake8, black, and isort
# prior to committing and will stop the commit if there are any
# warnings. Adjust BIN_DIR to the virtual environment where flake8,
# black, and isort are installed.
#
# To enable this hook, copy this file to ".git/hooks".

BIN_DIR=./.venv3/bin

CHANGED_FILES=$(git diff --cached --name-only --diff-filter=AM | grep .pyi || true)

if test -n "${CHANGED_FILES}" -a -d "${BIN_DIR}"; then
    ${BIN_DIR}/flake8 ${CHANGED_FILES}
    #${BIN_DIR}/black --check ${CHANGED_FILES}
    #${BIN_DIR}/isort --check-only ${CHANGED_FILES}

    # Replace the last two lines with the following lines
    # if you want to reformat changed files automatically
    # before committing. Please note that partial commits
    # (git add -p) will not work and will commit the whole
    # file!
    #
    ${BIN_DIR}/black ${CHANGED_FILES} || true
    ${BIN_DIR}/isort ${CHANGED_FILES} || true
    git add ${CHANGED_FILES}
fi
```

**Comment by blink1073** [comment]
Actually, it did fail on the files I changed, the check only looks at the stub files in CI.

File: pymongo-stubs/monitoring.pyi (hunk 4 above)
Comment (ShaneHarvey) [reply pymongo-stubs/monitoring.pyi:55]: Still need to update this service_id line.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

**Review by ShaneHarvey — APPROVED** [review]
LGTM! Please update the PR title+commit message to mention that you also added support for 3.12.

**Pull request merged**